R/mail.R

```diff
@@ -4,238 +4,566 @@
+#' @param end_year_month The year and month of the last file to be downloaded (format: 'YYYYMM').
+#' @param save_folder_path The folder path where all the downloaded mbox files will be stored.
+#' @param verbose if TRUE, prints detailed messages during the download process.
+#' @return Returns `save_folder_path`, the folder path where the mbox files are stored.
+#' @export
+download_mod_mbox <- function(mailing_list, start_year_month, end_year_month, save_folder_path, verbose = TRUE) {
+
+  ########## Extract Mailing List Name ##########
+  # Extract the mailing list name from the given URL. This is because the actual list name is
+  # embedded within the URL (after the 'list.html?').
+  # We are using 'sub()' to perform a simple string replacement, extracting everything after 'list.html?'.
+  mailing_list_name <- sub(".*list.html\\?(.+)", "\\1", mailing_list)
+  if (verbose) message("Base list extracted:", mailing_list_name, "\n")
+
+  ########## Prepare Year and Month ##########
+  # The start_year_month and end_year_month are in the format "YYYYMM".
+  # Split them into year and month for easier looping.
+  # Extract first 4 digits as start year, and last 2 digits as start month.
+  start_year <- as.numeric(substr(start_year_month, 1, 4))
+  start_month <- as.numeric(substr(start_year_month, 5, 6))
+  # Extract first 4 digits as end year, and last 2 digits as end month.
+  end_year <- as.numeric(substr(end_year_month, 1, 4))
+  end_month <- as.numeric(substr(end_year_month, 5, 6))
+
+  ########## Initialize Vectors for Failed Months ##########
+  # Vectors to track failed downloads.
+  failed_months <- character()
+
+  ########## Download Loop ##########
+  # Iterate over the years and months from start_year/month to end_year/month.
+  # This is done by looping over the years, and for each year, looping over the 12 months.
+  for (year in start_year:end_year) {
+    for (month in 1:12) {
+      # Skip months before the start_month or after the end_month for the start and end year.
+      if (year == start_year && month < start_month) next
+      if (year == end_year && month > end_month) break
+
+      ######### Construct URL and Save Path ##########
+      # Construct the month string (e.g., '2023-04') and the full download URL.
+      # Make sure the month has two digits.
+      month_str <- sprintf("%02d", month)
+      # Create a string in the format "YYYY-MM"
+      year_month_str <- sprintf("%04d-%02d", year, month)
+      # This constructs the URL from which the mbox for the current year and month will be downloaded.
+      # The format for the URL is fixed by Apache's Pony Mail service.
+      download_url <- stringi::stri_c("https://lists.apache.org/api/mbox.lua?list=", mailing_list_name, "&date=", year_month_str)
+
+      # Create the file name where the mbox will be saved locally, in the format ''kaiaulu_'YYYYMM.mbox'.
+      file_name <- stringi::stri_c("kaiaulu_", year, month_str, ".mbox")
```

Comment:
prefix hardcoded to kaiaulu. @daomcgill who is code reviewing this and executed your notebook?

Reply:
My current issue #317 has prevented me from doing a thorough review of Dao's notebook. I just posted my review and requested a few changes.